### PR TITLE
Virtual DOM implementation

### PR DESCRIPTION
**Pull request by jedimahdi**
There are multiple reasons which makes `snabbdom` unsuitable for the Monarch.

- It doesn't separate diffing and patching processes. So we cannot schedule these two different kind of tasks.
- The diffing process is recursive which at least comes with two drawbacks.
  - It blocks the main thread of JavaScript. So UI become unresponsive during the diff.
  - The process can't interrupt at some point.
- The virtual DOM tree is not extensible. So we cannot have custom nodes (e.g. thunk, async, suspense, etc...) easily.

BTW here we are with a new capable virtual DOM implementation which is highly inspired by the [elm/virtual-dom](https://github.com/elm/virtual-dom).

Essential features:

- The diffing and patching processes are completely separate. So can become scheduled differently.
- The diffing process is not recursive so it doesn't block the main thread and also it can interrupt if a new `VirtualDomTree` has been created.

### CODE REVIEW

**Review by the-dr-lazy — DISMISSED** [review]

**Comment by the-dr-lazy** [comment]
This PR is pending for #6

File: src/Monarch/VirtualDom/OutputHandlerTree.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Output handler tree ADT
+ */
+export type OutputHandlerTree = OutputHandlerTree.Root | OutputHandlerTree.Node
```
Comment (jedimahdi) [inline src/Monarch/VirtualDom/OutputHandlerTree.ts]: `OutputHandlerTree` data type is not a tree. It's just linked list of nodes that change the output from current node backward to the root.

Comment (the-dr-lazy) [reply src/Monarch/VirtualDom/OutputHandlerTree.ts]: sooriland_stickerpack_agadyvkaaj8smwu

Comment (the-dr-lazy) [reply src/Monarch/VirtualDom/OutputHandlerTree.ts]: Nice catch. It's kinda a `NonEmpty` but there is a difference between first element and others type. The first element is message dispatcher and should be of type `message -> unit` and others are message transformer and should be of type `messageA -> messageB`. So I think it's reasonable to introduce a new datatype for it.

File: src/Monarch/VirtualDom/Patch.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        case Patch.RemoveFromEnd:
+            return unsafe_applyRemoveFromEndPatch(domNode, patch)
+
+        case Patch.Append:
+            return unsafe_applyAppendPatch(domNode, patch)
+    }
+}
+
+function unsafe_applyTextPatch(textNode: Text, { text }: Patch.Text): void {
+    textNode.replaceData(0, text.length, text)
```
Comment (jedimahdi) [inline src/Monarch/VirtualDom/Patch.ts]: I think this line is wrong. Let's say we have
```js
const textNode = document.createTextNode("Hello World");
const newText = "text";
```
diff will create text patch and call this function
```js
textNode.replaceData(0, newText.length, newText);
```
and textNode will be "testo World"

Comment (the-dr-lazy) [reply src/Monarch/VirtualDom/Patch.ts]: Right. Let's play with old text.

**Review by the-dr-lazy — APPROVED** [review]
We did it.

**Comment by the-dr-lazy** [comment]
This PR closes #5.

**Pull request merged**